### ISSUE TITLE
Integrate accessibility checker within the page editor / inline preview

### ISSUE THREAD

**Issue by thibaudcolas** [wagtail/wagtail#10136]
## Is your proposal related to a problem?

The accessibility checker currently displays within the userbar. This means its output will only ever be seen if developers have enabled the userbar on the site, and when CMS users directly access a page on the site’s frontend.

## Describe the solution you'd like

Accessibility issues would be simpler to address if flagged as part of the page editing experience, alongside the live preview.

Potentially, we could also surface the issues when users click "Save draft" or "Publish" (like users would see normal form validation errors) – so it’s harder to miss the checks.

## Describe alternatives you've considered

Keep the checker to the userbar only.

## Additional context

This has been discussed in [Built-in accessibility checker for Wagtail](https://github.com/wagtail/wagtail/discussions/9262). We have existing designs for this produced by Nick.

There are a few critical considerations here. Architecture:

- Ideally the checker should work without the userbar being configured on the page – so it’d need to dynamically inject a script tag with Axe bundled.
- Since we display the results within the page editor, it would be much simpler to resolve any issues if they are connected to specific fields in the form. This is potentially very complex since the checker only reports on which part of the site’s _frontend_ the error is.

UX:

- It’s important for the rendering of any issues to not be too distracting, and not create confusion between form validation errors (mandatory to fix at this time) and accessibility checks failing (optional to fix)
- There isn’t enough space in the UI to display the editing form, the live preview, and checker results all at the same time. So it needs to be straightforward to move between the different panels.
- The UI should still account for errors that aren’t connected to any part of the form, either because they’re with content produced elsewhere, or because we couldn’t find which part of the form is relevant.
- Keyboard or screen reader users will also need a way to cycle between the different panels – matching the experience of the F6 key in Windows apps (so `Ctrl + F6` within a page?)

---

## Mapping accessibility errors to page fields

Here is early R&D, rule-by-rule. At a high level, matching by element text feels possible in a wide number of cases but not all. An alternative would be to change the rendering of Wagtail content on sites to contain field marks (repurposing comments’ `contentpath`?)

### empty-heading

* In rich text: Possible by relying either on Draftail block key, or text of rich text elements before/after.
* In Django fields: Impossible
* In StreamField: Impossible unless we introduced a special HeadingBlock as a CharBlock derivative as a CMS feature. For a project-specific HeadingBlock: at least likely to be a required field.

### heading-order

* In rich text: Possible by relying on Draftail block key, or heading text.
* In Django fields: Possible by text match.
* In StreamField: Possible by text match. Would be helped by a custom HeadingBlock.

### p-as-heading

* In rich text: Possible by relying on Draftail block key, or element text match.

### Empty button, link, frame

* Everywhere: likely too hard to do reliably.

### Table validation issues

* TableBlock: likely possible via text match, or could output a block key in our table template
* TypedTableBlock: perhaps possible via text match, or could output a block key in our table template

### Custom link text rule

* In rich text: Possible by relying on Draftail block key, or element text match.
* In Django fields: Possible by text match.
* In StreamField: Possible by text match.

### Custom image alt text rule

* In rich text: Possible by relying on Draftail block key, or alt text / other attributes match
* In Django fields: Possible by alt text / other attributes match
* In StreamField: Possible by alt text / other attributes match

**Comment by sande265** [comment]
Hi @thibaudcolas thibaudcolas I would like to work on this issue, it would be really nice if you assign this issue to me.

**Comment by laymonage** [comment]
Thanks for your interest @sande265! This is not a good issue for new contributors, as it requires in-depth knowledge about how the page editor is built and how fields and StreamField blocks are rendered. We also do not assign issues outside of the core team - we just ask that people indicate their intention to work on an issue to avoid duplicating efforts.

We recommend that you look for smaller issues, e.g. ones that are labelled as ["good first issue"](https://github.com/wagtail/wagtail/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22).

For further help in getting started, please see https://dev.to/lb/ten-tasty-ingredients-for-a-delicious-pull-request-hgc - you may also wish to join [our #new-contributors channel on Slack](https://github.com/wagtail/wagtail/wiki/Slack).

**Comment by thibaudcolas** [comment]
From @cnk in #10394:

> I want my content editors to make incremental improvements to my site's accessibility and mobile formatting. The accessibility checker is getting better and better - but at the moment it's warnings are only available from the userbar on live pages or when viewing the preview in a separate page. I would love for my users to be able to see accessibility warnings while they are using the mobile preview pane inside the page editor. The most consistent option is probably to add the userbar to the inline preview panels (perhaps only displaying the accessibility link), but there may be better options.
